File: bridge/BBBrowserExtensionInjector.ts
import type { BrowserLaunchOptions } from "./BrowserLauncher.js";
import { ExtensionInjector } from "./ExtensionInjector.js";

const DEFAULT_BROWSERBASE_BASE_URL = "https://api.browserbase.com";

function firstString(...values: unknown[]) {
  for (const value of values) {
    if (typeof value === "string" && value.trim()) return value.trim();
  }
  return null;
}

export class BBBrowserExtensionInjector extends ExtensionInjector {
  private extension_id: string | null = null;
  private zip_path: string | null = null;
  private cleanup: (() => Promise<void>) | null = null;

  async prepare() {
    const configured_extension_id = firstString(this.options.extension_id);
    if (configured_extension_id) {
      this.extension_id = configured_extension_id;
      return;
    }
    const extension_path = this.options.extension_path;
    if (!extension_path) return;
    this.zip_path = extension_path.endsWith(".zip") ? extension_path : await this.zipExtensionDir(extension_path);
    this.extension_id = await this.uploadExtension(this.zip_path);
  }

  getLauncherConfig(): BrowserLaunchOptions {
    if (!this.extension_id) return {};
    return { extension_id: this.extension_id };
  }

  async inject() {
    const discovered = await this.waitForReadyServiceWorker(this.options.service_worker_ready_timeout_ms ?? 60_000, {
      matched_only: this.options.trust_matched_service_worker,
    });
    return discovered ? { ...discovered, source: "bb" } : null;
  }

  async close() {
    await this.cleanup?.();
    this.cleanup = null;
  }

  private async zipExtensionDir(extension_path: string) {
    const [{ execFileSync }, fs, os, path] = await Promise.all([
      import("node:child_process"),
      import("node:fs"),
      import("node:os"),
      import("node:path"),
    ]);
    const zip_path = path.join(fs.mkdtempSync(path.join(os.tmpdir(), "modcdp-bb-extension-")), "extension.zip");
    execFileSync("zip", ["-X", "-qr", zip_path, "."], { cwd: extension_path });
    this.cleanup = async () => fs.rmSync(path.dirname(zip_path), { recursive: true, force: true });
    return zip_path;
  }

  private async uploadExtension(zip_path: string) {
    const browserbase_api_key = firstString(this.options.browserbase_api_key, process.env.BROWSERBASE_API_KEY);
    if (!browserbase_api_key) {
      throw new Error("BBBrowserExtensionInjector requires BROWSERBASE_API_KEY or launch.options.browserbase_api_key.");
    }
    const base_url = firstString(this.options.base_url, this.options.browserbase_base_url, process.env.BROWSERBASE_BASE_URL) ?? DEFAULT_BROWSERBASE_BASE_URL;
    const fs = await import("node:fs");
    const path = await import("node:path");
    const form = new FormData();
    form.append("file", new Blob([fs.readFileSync(zip_path)]), path.basename(zip_path));
    const response = await fetch(new URL("/v1/extensions", `${base_url.replace(/\/$/, "")}/`), {
      method: "POST",
      headers: { "X-BB-API-Key": browserbase_api_key },
      body: form,
    });
    if (!response.ok) {
      const text = await response.text().catch(() => "");
      throw new Error(`Browserbase POST /v1/extensions -> ${response.status}${text ? `: ${text}` : ""}`);
    }
    const extension = (await response.json()) as Record<string, unknown>;
    if (typeof extension.id !== "string" || !extension.id) {
      throw new Error(`Browserbase extension upload returned no id (got ${JSON.stringify(extension)})`);
    }
    return extension.id;
  }
}

File: client/go/BBBrowserExtensionInjector.go
package modcdp

import (
	"archive/zip"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
	"strings"
)

const DefaultBrowserbaseBaseURL = "https://api.browserbase.com"

type BBBrowserExtensionInjector struct {
	ExtensionInjector
	ExtensionID string
	ZipPath     string
	CleanupPath string
}

func NewBBBrowserExtensionInjector(options ExtensionInjectorConfig) BBBrowserExtensionInjector {
	return BBBrowserExtensionInjector{ExtensionInjector: NewExtensionInjector(options)}
}

func (i *BBBrowserExtensionInjector) Prepare() error {
	if i.Options.ExtensionID != "" {
		i.ExtensionID = i.Options.ExtensionID
		return nil
	}
	if i.ExtensionID != "" {
		return nil
	}
	extensionPath := i.Options.ExtensionPath
	if extensionPath == "" {
		return nil
	} else if strings.HasSuffix(extensionPath, ".zip") {
		i.ZipPath = extensionPath
	} else {
		zipPath, cleanupPath, err := zipExtensionDir(extensionPath)
		if err != nil {
			return err
		}
		i.ZipPath = zipPath
		i.CleanupPath = cleanupPath
	}
	extensionID, err := i.uploadExtension(i.ZipPath)
	if err != nil {
		_ = i.Close()
		return err
	}
	i.ExtensionID = extensionID
	return nil
}

func (i *BBBrowserExtensionInjector) GetLauncherConfig() LaunchOptions {
	if i.ExtensionID == "" {
		return LaunchOptions{}
	}
	return LaunchOptions{ExtensionID: i.ExtensionID}
}

func (i *BBBrowserExtensionInjector) Inject() (*ExtensionInjectionResult, error) {
	extensionID := i.Options.ExtensionID
	i.Options.ExtensionID = ""
	defer func() { i.Options.ExtensionID = extensionID }()

	discovered, err := i.waitForReadyServiceWorker(i.Options.ServiceWorkerReadyTimeoutMS, i.Options.TrustMatchedServiceWorker)
	if err != nil || discovered == nil {
		return discovered, err
	}
	discovered.Source = "bb"
	return discovered, nil
}

func (i *BBBrowserExtensionInjector) Close() error {
	if i.CleanupPath != "" {
		_ = os.RemoveAll(i.CleanupPath)
		i.CleanupPath = ""
	}
	return nil
}

func (i *BBBrowserExtensionInjector) uploadExtension(zipPath string) (string, error) {
	browserbaseAPIKey := firstNonEmptyString(i.Options.BrowserbaseAPIKey, os.Getenv("BROWSERBASE_API_KEY"))
	if browserbaseAPIKey == "" {
		return "", fmt.Errorf("BBBrowserExtensionInjector requires BROWSERBASE_API_KEY or launch.options.browserbase_api_key")
	}
	baseURL := firstNonEmptyString(i.Options.BaseURL, i.Options.BrowserbaseBaseURL, os.Getenv("BROWSERBASE_BASE_URL"), DefaultBrowserbaseBaseURL)
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)
	fileWriter, err := writer.CreateFormFile("file", filepath.Base(zipPath))
	if err != nil {
		return "", err
	}
	file, err := os.Open(zipPath)
	if err != nil {
		return "", err
	}
	if _, err := io.Copy(fileWriter, file); err != nil {
		_ = file.Close()
		return "", err
	}
	if err := file.Close(); err != nil {
		return "", err
	}
	if err := writer.Close(); err != nil {
		return "", err
	}
	request, err := http.NewRequest("POST", strings.TrimRight(baseURL, "/")+"/v1/extensions", body)
	if err != nil {
		return "", err
	}
	request.Header.Set("X-BB-API-Key", browserbaseAPIKey)
	request.Header.Set("Content-Type", writer.FormDataContentType())
	response, err := http.DefaultClient.Do(request)
	if err != nil {
		return "", err
	}
	defer response.Body.Close()
	responseBody, _ := io.ReadAll(response.Body)
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return "", fmt.Errorf("Browserbase POST /v1/extensions -> %d%s", response.StatusCode, formatResponseBody(responseBody))
	}
	var payload map[string]any
	if err := json.Unmarshal(responseBody, &payload); err != nil {
		return "", err
	}
	extensionID, _ := payload["id"].(string)
	if extensionID == "" {
		return "", fmt.Errorf("Browserbase extension upload returned no id (got %s)", responseBody)
	}
	return extensionID, nil
}

func zipExtensionDir(extensionPath string) (string, string, error) {
	dir, err := os.MkdirTemp("", "modcdp-bb-extension-")
	if err != nil {
		return "", "", err
	}
	zipPath := filepath.Join(dir, "extension.zip")
	zipFile, err := os.Create(zipPath)
	if err != nil {
		_ = os.RemoveAll(dir)
		return "", "", err
	}
	archive := zip.NewWriter(zipFile)
	walkErr := filepath.WalkDir(extensionPath, func(path string, entry os.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if entry.IsDir() {
			return nil
		}
		relative, err := filepath.Rel(extensionPath, path)
		if err != nil {
			return err
		}
		writer, err := archive.Create(relative)
		if err != nil {
			return err
		}
		file, err := os.Open(path)
		if err != nil {
			return err
		}
		defer file.Close()
		_, err = io.Copy(writer, file)
		return err
	})
	closeArchiveErr := archive.Close()
	closeFileErr := zipFile.Close()
	if walkErr != nil || closeArchiveErr != nil || closeFileErr != nil {
		_ = os.RemoveAll(dir)
		if walkErr != nil {
			return "", "", walkErr
		}
		if closeArchiveErr != nil {
			return "", "", closeArchiveErr
		}
		return "", "", closeFileErr
	}
	return zipPath, dir, nil
}

func formatResponseBody(body []byte) string {
	if len(body) == 0 {
		return ""
	}
	return ": " + string(body)
}
